docs/blog/2024-08-27-challenges-of-microservices.md

```diff
@@ -0,0 +1,152 @@
+
+:::info
+Apache Kafka is my favorite distributed system. In Apache Kafka, the _Controller_ is a special Kafka server that is responsible for deciding which partition replicas are hosted on (and led by) which brokers. If the broker who was in charge of a partition goes down, then the Controller chooses a new broker from the ISR to take its place.
+
+Therefore, the _Controller_ in Apache Kafka can be thought of as a _leader_.
+:::
+
+While systems like Apache Kafka have clear leaders (for example, the _Controller_ may re-assign partition leadership if the cluster becomes too imbalanced), in a microservice-based system there is no central leader to ensure that the chips fall correctly. 
+
+You can think of our e-commerce microservice flow as a line of dominoes falling. Once the process starts, no one entity is responsible for ensuring its completion. The business workflow moves from `orders` to `inventory` to `payments` and so on. If `payments` fails for some reason (perhaps a network partition makes the Stripe API unavailable), then it's quite possible that the `shipping` service never finds out about the workflow.
+
+However, in real life such outcomes are not acceptable. This means that every single player in the system must:
+
+1. Have built-in reliability mechanisms.
+2. Understand the preceding and subsequent steps of the business process to route traffic.
+
+## The Challenges
+
+So far, we have established that there are many players involved in a business process, yet there's no one player in charge of ensuring that the trains run on time. This yields three problems:
+
+1. **Reliability** in the face of infrastructure failures.
+2. **Observability** to enable system optimization and debugging.
+3. **Coupling** of microservices to each other makes it hard to modify the system in response to new business requirements.
+
+### Reliability and Correctness
+
+Processing orders is a mission-critical use-case. This means that orders should always complete and never be dropped (for example, we shouldn ever charge the customer's credit card and not ship the product to them).
+
+However, asynchronous processing such as that which I outlined above is prone to failures. For example, if you chain microservices together with direct RPC calls, a single network partition can cause an order to get stuck. Even with a reliable message broker such as Apache Kafka or AWS SQS sitting between your microservices, a write to the message broker could fail _after_ the payment went through, still resulting in a stuck order.
+
+Just as communication _between_ microservices can fail, the actions performed _by_each microservice can also fail. If the Stripe API is down, or if the credit card is invalid, we can't just stop processing the order there! We must notify the customer of what went wrong and also release the inventory that we reserved.
```

Comment:
Highlight the call out to external systems.  
Suggestion:
Just as communication _between_ microservices can fail, the actions performed _by_each microservice can also fail.  In many cases actions performed by a microservice must be performed and the resulting return determines what steps to take next.  External systems, due to the additional steps involved in communicating them with, are often more prone for failure.  For example if the Stripe API is down, or if the credit card is invalid, we can't just stop processing the order there! We must notify the customer of what went wrong and also release the inventory that we reserved.